core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java

```diff
@@ -934,6 +944,28 @@ public static DistributedStagesScheduler create(
             return distributedStagesScheduler;
         }
 
+        private static int getEstimatedHashPartitionCount(StageManager stageManager, int maxHashPartitionCount)
+        {
+            List<PartitioningScheme> schemes = stageManager.getStagesInTopologicalOrder().stream()
+                    .map(stage -> stage.getFragment().getPartitioningScheme())
+                    .filter(schema -> !SINGLE_PARTITIONING_HANDLES.contains(schema.getPartitioning().getHandle()))
+                    .collect(toImmutableList());
+
+            if (schemes.stream().allMatch(scheme -> scheme.getHashPartitionCount().isEmpty())) {
+                // We expect all to be empty when DetermineHashPartitionCount optimizer rule is off
+                return maxHashPartitionCount;
+            }
+
+            // We expect all should have hashPartitionCount when DetermineHashPartitionCount optimizer rule is on
```

Comment:
it seems to be fragile when plan contains both system partitioning handles and connector partitioning handles